### PR TITLE
Initial Refactor Gateway Allocator

### PR DESCRIPTION
**Pull request by kimchy**
Break it into more manageable code by separating allocation primaries and allocating replicas. Start adding basic unit tests for primary shard allocator.

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/gateway/PrimaryShardAllocator.java [hunk 1]
```diff
@@ -0,0 +1,298 @@
...
+            } catch (Exception e) {
+                logger.warn("[{}][{}] failed to derived initial_shards from value {}, ignore allocation for {}", shard.index(), shard.id(), initialShards, shard);
+            }
+        }
+
+        return nodesAndVersions.allocationsFound >= requiredAllocation;
+    }
+
+    /**
+     * Based on teh nodes and versions, build the list of yes/no/throttle nodes that the shard applies to.
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/gateway/PrimaryShardAllocator.java]: typo

File: core/src/test/java/org/elasticsearch/gateway/PrimaryShardAllocatorTests.java [hunk 2]
```diff
@@ -0,0 +1,300 @@
...
+        assertThat(allocation.routingNodes().ignoredUnassigned().get(0).shardId(), equalTo(shardId));
+    }
+
+    /**
+     * Tests when the node returns that no data was found for it (-1), it will be moved to ignore unassigned.
+     */
+    @Test
+    public void testNoAllocationFound() {
+        RoutingAllocation allocation = setupOneNodeOneShard(yesAllocationDeciders());
+        testAllocator.addData(node1, -1);
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/gateway/PrimaryShardAllocatorTests.java]: can we randomize between a store exception and the current code? (alternatively add another test)

File: core/src/test/java/org/elasticsearch/gateway/PrimaryShardAllocatorTests.java [hunk 3]
```diff
@@ -0,0 +1,300 @@
...
+        allocation = new RoutingAllocation(yesAllocationDeciders(), state.routingNodes(), state.nodes(), null);
+        changed = testAllocator.allocateUnassigned(allocation);
+        assertThat(changed, equalTo(true));
+        assertThat(allocation.routingNodes().ignoredUnassigned().size(), equalTo(0));
+        assertThat(allocation.routingNodes().shardsWithState(ShardRoutingState.UNASSIGNED).size(), equalTo(2)); // replicas
+        assertThat(allocation.routingNodes().shardsWithState(ShardRoutingState.INITIALIZING).size(), equalTo(1));
+        assertThat(allocation.routingNodes().shardsWithState(ShardRoutingState.INITIALIZING).get(0).currentNodeId(), equalTo(node2.id()));
+    }
+
+    private RoutingAllocation setupOneNodeOneShard(AllocationDeciders deciders) {
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/gateway/PrimaryShardAllocatorTests.java]: nit: we actually setup  3 nodes, call this routingAllocationWithOneShard?

File: core/src/main/java/org/elasticsearch/gateway/PrimaryShardAllocator.java [hunk 4]
```diff
@@ -0,0 +1,298 @@
...
+        }
+        return new NodesToAllocate(Collections.unmodifiableList(yesNodes), Collections.unmodifiableList(throttledNodes), Collections.unmodifiableList(noNodes));
+    }
+
+    /**
+     * Builds a list of nodes and version
+     */
+    private NodesAndVersions buildNodesAndVersions(ShardRouting shard, boolean recoveryOnAnyNode, Set<String> ignoreNodes,
+                                                   AsyncShardFetch.FetchResult<TransportNodesListGatewayStartedShards.NodeGatewayStartedShards> shardState) {
+        final Map<DiscoveryNode, Long> nodesWithVersion = Maps.newHashMap();
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/gateway/PrimaryShardAllocator.java]: we only need an array here - we don't do anything with the version and we copy it into one anyway:  https://github.com/elastic/elasticsearch/pull/12335/files#diff-ad5388a03f5e080b452190f4eb47f33aR244

Might as well use an arraylist from the beginning.

Comment (bleskes) [inline core/src/main/java/org/elasticsearch/gateway/PrimaryShardAllocator.java]: Miss the sorting based on the version.  Sorry for the noise..

File: core/src/test/java/org/elasticsearch/gateway/PrimaryShardAllocatorTests.java (hunk 3 above)
Comment (kimchy) [inline core/src/test/java/org/elasticsearch/gateway/PrimaryShardAllocatorTests.java]: will do

File: core/src/test/java/org/elasticsearch/gateway/PrimaryShardAllocatorTests.java (hunk 2 above)
Comment (kimchy) [inline core/src/test/java/org/elasticsearch/gateway/PrimaryShardAllocatorTests.java]: sure

File: core/src/main/java/org/elasticsearch/gateway/PrimaryShardAllocator.java (hunk 1 above)
Comment (kimchy) [inline core/src/main/java/org/elasticsearch/gateway/PrimaryShardAllocator.java]: belh

File: core/src/main/java/org/elasticsearch/gateway/GatewayAllocator.java [hunk 5]
```diff
@@ -132,430 +112,85 @@ public void applyFailedShards(FailedRerouteAllocation allocation) {
...
-            AsyncShardFetch.FetchResult<TransportNodesListGatewayStartedShards.NodeGatewayStartedShards> shardState = fetch.fetchData(nodes, metaData, allocation.getIgnoreNodes(shard.shardId()));
-            if (shardState.hasData() == false) {
-                logger.trace("{}: ignoring allocation, still fetching shard started state", shard);
-                unassignedIterator.remove();
-                routingNodes.ignoredUnassigned().add(shard);
-                continue;
-            }
-            shardState.processAllocation(allocation);
+        changed |= primaryShardAllocator.allocateUnassigned(allocation);
+        changed |= replicaShardAllocator.allocateUnassigned(allocation);
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/gateway/GatewayAllocator.java]: can we add an assertion going out that going out of the primary shard allocator we don't have any primary shards in the unassigned list, unless we expect them to be there? (primaryAllocatedPostApi is false or restoreSource != null)

**Comment by bleskes** [comment]
LGTM. Left one assertion ask and minor comments. Nice breakage - much easier to test.

File: core/src/main/java/org/elasticsearch/gateway/GatewayAllocator.java (hunk 5 above)
Comment (kimchy) [inline core/src/main/java/org/elasticsearch/gateway/GatewayAllocator.java]: I will do the assert when I remove the primaryAllocated flag in a different change

Comment (bleskes) [inline core/src/main/java/org/elasticsearch/gateway/GatewayAllocator.java]: OK.

**Comment by kimchy** [comment]
pushed, not sure why it didn't pick up on merged

**Pull request closed without merging**